src/io_ops/nonblocking_io.rs

```diff
@@ -0,0 +1,103 @@
+use super::FileDescriptorRawHandle;
+use async_trait::async_trait;
+use libc::O_NONBLOCK;
+use libc::{self, c_int, c_void, size_t};
+use log::error;
+use std::io;
+use std::sync::Arc;
+use tokio::io::unix::AsyncFd;
+
+#[derive(Debug, Clone)]
+pub struct SubChannel {
+    fd: Arc<AsyncFd<FileDescriptorRawHandle>>,
+    shared: bool, // If its a shared file handle when asked to close, noop.
```

Comment:
Sg, i think Arc should be free to use/access outside of drop/clone, and not involve any indirection, so either using a single Arc wrapping each value or shared Arc's should achieve the same purpose and be much simpler. Will update.

Reply:
Looking at this more, there is a little bit of awkwardness that we want to be able to call close -- but we won't know everything has dropped. (Once FD closes, so close the others to make it all shutdown). I've refactored it now i think to be more clear, use Arc + Drop, but also for the inner FD we do have a close method that synchronizes using an atomic boolean to ensure we only call it once.

Reply:
FWIW as part of my work removing as much unsafe as possible I've been replacing the raw fd handing (and calls to libc::write, libc::close, etc.) with a normal `std::fs::File`.  That way we have confidence that things are only closed once and there are no data races because such things are impossible without unsafe.

Reply:
Is this a safe/guaranteed operation for fuse? I had wondered when looking at this, but I couldn't find any documentation on either side saying the atomic operation's that fuse uses in the kernel would be safe/respected going via the file API. This code relies on using iovec writes which are guaranteed atomic in the fuse kernel handling. Will the std::fs::File API make the same guaranteed?(Wondering if stuff winds up getting chunked, written in smaller/multiple writes it could get interleaved or not as one fuse operation?)

Reply:
Well the docs for [write](https://doc.rust-lang.org/std/io/trait.Write.html#tymethod.write) say:

> A call to write represents at most one attempt to write to any wrapped object.

and that seems to carry over to [write_vectored](https://doc.rust-lang.org/std/io/trait.Write.html#method.write_vectored):

> This method must behave as a call to write with the buffers concatenated would.

And the impl works as you'd expect/want: https://github.com/rust-lang/rust/blob/a15f484b918a4533ad633ea903ccce82910af342/library/std/src/sys/unix/fd.rs#L151-L160

So I think it's pretty safe.

Reply:
Unless I’m mistaken, I think your taking some of these guarantee’s more strongly than they are intended. — write_vectored doesn’t guarantee that there is one underlying write call — indeed it has a default fall back to call write on the first non-empty buffer (which would be an invalid write in our use case here). There is a nightly API to query if the write would be vectored. This would corrupt the fuse operations afaik. 

(One attempt doesn’t necessarily mean one system call no? It means if there’s a failure or partial write at the OS level it won’t attempt to call write again on that and instead would return Err or Ok(num bytes) processed. Since we explicitly require the system call parity i don’t think these are quite the same strength of a contract. )


I guess for these internal ones where the fuse kernel is abusing the file system API, I’m pretty slow to move to a higher level API not designed for this. The current implementations for Unix look right to me (though we pass the FD around more and explicitly call close to terminate operations which isn’t supported in the std::fs::File api), but I believe they could remain within their contract and break fuse in subtle ways(iovec write being an example).

Reply:
> we pass the FD around more and explicitly call close to terminate operations which isn’t supported in the std::fs::File api

This was one of my motivations for moving to the `File` API. I was working through the fuse_sys code and found a few places where it looked like we could `close()` twice or perform operations after the close. By using the `File` API and removing the `unsafe` blocks we'd be forced to get this right.

The exception is when linking against libfuse3 where the FD is by rights owned by the libfuse session, but my plan there would be to do `File::into_raw_fd` before destroying the session.

> where the fuse kernel is abusing the file system API

Fuse isn't that special in this regard.  The concerns about atomic writes would apply to named pipes just the same (`PIPE_BUF`)

> I believe they could remain within their contract and break fuse in subtle ways (iovec write being an example).

Rust being a systems language I think this is safe to assume that a single call to `write_vectored` results in a single call to `writev`. The Rust maintainers take backwards compatibility very seriously, Changing this behaviour would break users that reasonably rely on it, as such would be a compatibility break.

Reply:
`write_vectored` calling `writev` is explicitly not in the file API though as an example. The current implementation on Unix does this. But the API does not require/and by default doesn’t do this.

(there’s also the requirement that you have a mutable handle on the File in order to read/write from it in the std::fs::File api that I’m aware which wouldn’t mix well with the current API where its presumed that users can use the Reply passed in from another thread — unless we wrap it in mutex locks?)

Reply:
Anyway I’m relatively new to this library so I’d defer to @cberner as the maintainer, this is a low level library, so I’d be inclined to say it should aim to use the bare metal API if its at all ambiguous. Limiting the access surface of the unsafe/using Arc’s and drop behaviors should all have i believe the same effect as a reduced surface File API.

Reply:
Even as the maintainer, I'm only medium familiar with this part of the code, since I took it over from the original author :)

I'd like to have as little `unsafe` code as possible. For the specific case of `write_vectored` though, I think we should stick with `writev` until there's a way to check that `write_vectored` is atomic. Debugging problems from a non-atomic `write_vectored` implementation would be frustrating